backends/tfhe-cuda-backend/cuda/src/crypto/keyswitch.cuh

```diff
@@ -146,12 +147,169 @@ void execute_keyswitch_async(cudaStream_t *streams, uint32_t *gpu_indexes,
+
+  // Iterate through all lwe elements
+  for (int i = 0; i < lwe_dimension_in; i++) {
+    // Round and prepare decomposition
+    Torus a_i = round_to_closest_multiple(lwe_in[i], base_log, level_count);
+
+    Torus state = a_i >> (sizeof(Torus) * 8 - base_log * level_count);
+    Torus mod_b_mask = (1ll << base_log) - 1ll;
+
+    // block of key for current lwe coefficient (cur_input_lwe[i])
+    auto ksk_block = &fp_ksk[i * ksk_block_size];
+    for (int j = 0; j < level_count; j++) {
+      auto ksk_glwe = &ksk_block[j * glwe_size * polynomial_size];
+      // Iterate through each level and multiply by the ksk piece
+      auto ksk_glwe_chunk = &ksk_glwe[chunk_id * coef_per_block];
+      Torus decomposed = decompose_one<Torus>(state, mod_b_mask, base_log);
+      local_glwe_chunk[shmem_index] -= decomposed * ksk_glwe_chunk[shmem_index];
+    }
+  }
+
+  // Persist
+  glwe_out[tid] = local_glwe_chunk[shmem_index];
+}
+
+// public functional packing keyswitch
+//
+// blockIdx.y - input
+// chunk_count = glwe_size * polynomial_size / threads.
+template <typename Torus>
+__global__ void
+fp_keyswitch_lwe_list_to_glwe(Torus *glwe_array_out, Torus *lwe_array_in,
+                              Torus *fp_ksk, uint32_t lwe_dimension_in,
+                              uint32_t glwe_dimension, uint32_t polynomial_size,
+                              uint32_t base_log, uint32_t level_count,
+                              Torus *d_mem) {
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  const int glwe_accumulator_size = (glwe_dimension + 1) * polynomial_size;
+  const int lwe_size = (lwe_dimension_in + 1);
+
+  const int input_id = blockIdx.y;
+  const int degree = input_id;
+
+  // Select an input
+  auto lwe_in = lwe_array_in + input_id * lwe_size;
+  auto ks_glwe_out = d_mem + input_id * glwe_accumulator_size;
+  auto glwe_out = glwe_array_out + input_id * glwe_accumulator_size;
+  // KS LWE to GLWE
+  keyswitch_lwe_ciphertext_into_glwe_ciphertext(
+      ks_glwe_out, lwe_in, fp_ksk, lwe_dimension_in, glwe_dimension,
+      polynomial_size, base_log, level_count);
+
+  // P * x ^degree
+  auto in_poly = ks_glwe_out + (tid / polynomial_size) * polynomial_size;
+  auto out_result = glwe_out + (tid / polynomial_size) * polynomial_size;
+  polynomial_accumulate_monic_monomial_mul(out_result, in_poly, degree,
+                                           tid % polynomial_size,
+                                           polynomial_size, 1, true);
+}
+
+/// To-do: Rewrite this kernel for efficiency
+template <typename Torus>
+__global__ void accumulate_glwes(Torus *glwe_out, Torus *glwe_array_in,
+                                 uint32_t glwe_dimension,
+                                 uint32_t polynomial_size, uint32_t num_lwes) {
+  const int tid = threadIdx.x + blockIdx.x * blockDim.x;
+
+  extern __shared__ int8_t sharedmem[];
+  const int shmem_index = threadIdx.x;
```

Comment:
Again  here I'm not sure using shared memory helps, as you don't need to synchronize the thread block.